memcached.c

```diff
@@ -5805,6 +5805,7 @@ static void drive_machine(conn *c) {
                             if (settings.verbose) {
                                 fprintf(stderr, "SSL connection failed with error code : %d : %s\n", err, strerror(errno));
                             }
+                            SSL_free(ssl);
```

Comment:
+1 @dormando.
This should be :+1: 
`
if (ssl) {
                                SSL_shutdown(ssl);
                                SSL_free(ssl);
                            }
`
Also in the `conn_free`, probably better as one static method in the `tls.c` and used in all 3 places.

Reply:
if (ssl == NULL) check is directly above so there shouldn't need to be another check. Looking at the docs I'm not sure SSL_shutdown is needed either, but maybe it can't hurt? Also a function call won't help as there's no connection structure in this case :)

Reply:
Yes, null check is repetitive. We have to do this in the `conn_free` as well. A method call can be something like `SSL_close(SSL *ssl)` as we always have a `SSL` pointer.

Reply:
Right, `SSL_shutdown` is not needed here.
From the man pages:
`Note that SSL_shutdown() must not be called if a previous fatal error has occurred on a connection i.e. if SSL_get_error() has returned SSL_ERROR_SYSCALL or SSL_ERROR_SSL.`, which is the case here. Only needed in `conn_free`. No need of another method.

Reply:
Cool yeah. I think this is good as is then. Will wait for the reporter to test it before merging, but not planning on making a release right this moment anyway.

Reply:
Yes. Don't we need the same in `conn_free` method? If so we need `SSL_shutdown` there.

Reply:
I think so yeah. It does call conn_free() with an SSL context. I'll update this patch to cover it. Since it's a malloc failure I'm dubious that code's ever been called though...

conn_new() for a client connection is only ever called in thread.c, which checks for a null response and closes the client socket. Check either goes there or in conn_free(). I'm leaning toward the dispatch call since that already owns closing the socket, though that feels a bit off.

Reply:
Yeah I'll add it in the dispatch handler. We also get a null return if event_add() fails for some reason, which doesn't call conn_free().